src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+
+    def containsObject(self, obj):
+        # TODO improve this procedure!
+        # Fast check
+        for c in obj.corners:
+            if not self.containsPoint(c):
+                return False
+        # Slow check
+        gps = [self.pointToGrid(corner) for corner in obj.corners]
+        x, y = zip(*gps)
+        minx, maxx = findMinMax(x)
+        miny, maxy = findMinMax(y)
+        for x in range(minx, maxx+1):
+            for y in range(miny, maxy+1):
+                p = self.gridToPoint((x, y))
+                if self.grid[y, x] == 1 and obj.containsPoint(p):
+                    return False
+        return True
+
+    def containsRegionInner(self, reg, tolerance):
+        raise NotImplementedError
+
+###################################################################################################
+# View Regions
+###################################################################################################
+
+class DefaultViewRegion(MeshVolumeRegion):
+    """ The default view region shape.
+
+    The default view region can take several forms, depending on the viewAngles parameter:
+
+    * Case 1:       viewAngles[0] = 360 degrees
+
+      * Case 1.a:   viewAngles[1] = 180 degrees         => Full Sphere View Region
+      * Case 1.b:   viewAngles[1] < 180 degrees         => Sphere - (Cone + Cone) (Cones on z axis expanding from origin)
+
+    * Case 2:       viewAngles[0] = 180 degrees
+
+      * Case 2.a:   viewAngles[1] = 180 degrees         => Hemisphere View Region
+      * Case 2.b:   viewAngles[1] < 180 degrees         => Hemisphere - (Cone + Cone) (Cones on appropriate hemispheres)
+
+    * Case 3:       viewAngles[1] = 180 degrees
+
+      * Case 3.a:   viewAngles[0] < 180 degrees         => Sphere intersected with Pyramid View Region
+      * Case 3.b:   viewAngles[0] > 180 degrees         => Sphere - Backwards Pyramid View Region
+
+    * Case 4:       viewAngles[0 & 1] < 180             => Capped Pyramid View Region
+    * Case 5:       viewAngles[0] > 180, Altitude < 180 => (Sphere - (Cone + Cone) (Cones on appropriate hemispheres)) - Backwards Capped Pyramid View Region
+
+    Args:
+        visibleDistance: The view distance for this region.
+        viewAngles: The view angles for this region.
+        name: An optional name to help with debugging.
+        position: An optional position, which determines where the center of the region will be.
+        rotation: An optional Orientation object which determines the rotation of the object in space.
+        orientation: An optional vector field describing the preferred orientation at every point in
+          the region.
+        tolerance: Tolerance for collision computations.
+    """
+    def __init__(self, visibleDistance, viewAngles, name=None, position=Vector(0,0,0), rotation=None,\
+        orientation=None, tolerance=1e-8):
+        # Bound viewAngles from either side.
+        viewAngles = tuple([min(angle, math.tau) for angle in viewAngles])
```

Comment:
Either we should disallow `viewAngles[1] > 180 deg` (raising an exception), or we should handle it (which would require more cases).